### PR TITLE
Apply `Indexby` after loading relations

### PR DESCRIPTION
**Pull request by SamMousa**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | maybe
| Fixed issues  | #9342

### CODE REVIEW

File: framework/web/Cookie.php [hunk 1]
```diff
@@ -65,4 +67,35 @@ public function __toString()
     {
         return (string) $this->value;
     }
+
+    /**
+     * Unserializes a cookie received by the client.
+     * @param string $data
+     * @return self|null
+     */
+    public static function fromDataString($data)
```
Comment (samdark) [inline framework/web/Cookie.php]: Doesn't look like it belongs to this pull request.

File: framework/web/Request.php [hunk 2]
```diff
@@ -1286,13 +1286,9 @@ protected function loadCookies()
...
                 }
-                $data = @unserialize($data);
-                if (is_array($data) && isset($data[0], $data[1]) && $data[0] === $name) {
-                    $cookies[$name] = new Cookie([
-                        'name' => $name,
-                        'value' => $data[1],
-                        'expire' => null,
-                    ]);
+
+                if (($cookie = Cookie::fromDataString($data)) !== null && $cookie->name === $name) {
```
Comment (samdark) [inline framework/web/Request.php]: Same here.

**Review by samdark — CHANGES_REQUESTED** [review]

**Comment by SamMousa** [comment]
Hmm based it off the wrong branch.

File: framework/db/ActiveQueryTrait.php [hunk 3]
```diff
@@ -111,45 +111,20 @@ public function with()
...
-                if (is_string($this->indexBy)) {
-                    $key = $row[$this->indexBy];
-                } else {
-                    $key = call_user_func($this->indexBy, $row);
-                }
-                $models[$key] = $row;
-            }
+            return $rows;
         } else {
+            $models = [];
```
Comment (SilverFire) [inline framework/db/ActiveQueryTrait.php]: No need for `else` above

**Review by SilverFire — CHANGES_REQUESTED** [review]

File: framework/db/ActiveQueryTrait.php (hunk 3 above)
Comment (SamMousa) [inline framework/db/ActiveQueryTrait.php]: I know there's no need but I find it more readable, early return statements tend to make your code less readable; is it in violation of the coding  standard?

Comment (SilverFire) [inline framework/db/ActiveQueryTrait.php]: Hm, personally for me early return statements are more readable :)

**Comment by samdark** [comment]
@SamMousa would you please fix the branching?

File: framework/db/ActiveQueryTrait.php (hunk 3 above)
Comment (arogachev) [inline framework/db/ActiveQueryTrait.php]: It's documented in Core code style guide, [Control statements section](https://github.com/yiisoft/yii2/blob/master/docs/internals/core-code-style.md#54-control-statements):

> Prefer avoiding `else` after `return` where it makes sense. Use [guard conditions](http://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html).
> 
> ```php
> $result = $this->getResult();
> if (empty($result)) {
>     return true;
> } else {
>     // process result
> }
> ```
> 
> is better as
> 
> ```php
> $result = $this->getResult();
> if (empty($result)) {
>    return true;
> }
> 
> // process result
> ```

**Comment by SamMousa** [comment]
Yes, it's on my list

**Comment by SamMousa** [comment]
@samdark I've reverted the erroneous commit.

File: tests/framework/db/ActiveRecordTest.php [hunk 4]
```diff
@@ -1439,4 +1439,20 @@ public function testUnlinkAllOnConditionViaTable()
...
+
+    /**
+     * https://github.com/yiisoft/yii2/pull/13891
+     */
+    public function testIndexByAfterLoadingRelations()
+    {
+        $orderClass = $this->getOrderClass();
+
+        $this->assertNotEmpty($orderClass::find()->with('customer')->indexBy(function($order) {
+            return isset($order->product->id) ? $order->product->id : null;
```
Comment (Faryshta) [inline tests/framework/db/ActiveRecordTest.php]: this is where you should call the `assertNotEmpty()` method.

File: tests/framework/db/ActiveRecordTest.php [hunk 5]
```diff
@@ -1439,4 +1439,20 @@ public function testUnlinkAllOnConditionViaTable()
...
     }
+
+    /**
+     * https://github.com/yiisoft/yii2/pull/13891
+     */
+    public function testIndexByAfterLoadingRelations()
+    {
+        $orderClass = $this->getOrderClass();
+
+        $this->assertNotEmpty($orderClass::find()->with('customer')->indexBy(function($order) {
```
Comment (Faryshta) [inline tests/framework/db/ActiveRecordTest.php]: why does the `with()` statement calls the customer relation and the index uses the product relation?

File: tests/framework/db/ActiveRecordTest.php [hunk 6]
```diff
@@ -1439,4 +1439,20 @@ public function testUnlinkAllOnConditionViaTable()
...
+            return isset($order->product->id) ? $order->product->id : null;
+        })->all());
+
+        $this->assertNotEmpty($orderClass::find()->with('customer')->indexBy('product.id')->all());
```
Comment (Faryshta) [inline tests/framework/db/ActiveRecordTest.php]: maybe here you should call https://phpunit.de/manual/current/en/appendixes.assertions.html#appendixes.assertions.assertArrayHasKey

**Comment by samdark** [comment]
Seems tests are failing...

**Comment by SilverFire** [comment]
Thank you!

**Review by cebe — COMMENTED** [review]
Looks good to me!

**Comment by klimov-paul** [comment]
This PR is BC break: NOSQL ActiveQuery classes are broken because of it:

- [mongodb](https://github.com/yiisoft/yii2-mongodb/issues/251)
- [filedb](https://github.com/yii2tech/filedb/issues/10)

I suspect the same problem at `yii2-elasticsearch` repository too. Someone should check it out and apply a fix if necessary.

**Comment by klimov-paul** [comment]
`yii2-redis` should be checked as well.

**Comment by cebe** [comment]
yes, yii2-redis is broken too: https://github.com/yiisoft/yii2-redis/issues/141

**Comment by cebe** [comment]
elasticsearch seems not affected

**Pull request merged**